### ISSUE TITLE
Implement For Each on the Dictionary

### ISSUE THREAD

**Issue by dimitrimitropulos** [VBA-tools/VBA-Dictionary#17]
observe the following comparison:

```
Sub implementForEach()
    Dim x As New Dictionary
    Dim y As New Scripting.Dictionary
    
    x.Add "a", 1
    y.Add "a", 1
    
    Dim z As Variant
    
    ' Run-time error '438': Object doesn't support this property or method
    For Each z In x
        Debug.Print z
    Next
    
    For Each z In y
        Debug.Print z 'prints the key
    Next

End Sub
```

The current implementation provides a way to For Each using `.Keys` and `.Items`, but not with the dictionary object itself (as Scripting.Dictionary allows).

**Comment by dimitrimitropulos** [comment]
The solution to this might look something like:
```
Public Function NewEnum() As IUnknown 'or IEnumVARIANT
#If Mac Or Not UseScriptingDictionaryIfAvailable Then
    Set NewEnum = dict_pKeyValues.[_NewEnum]
#Else
    Set NewEnum = dict_pDictionary
#End If
End Function
```
but I can't quite get the syntax to cooperate.

**Comment by timhall** [comment]
Wow, can't believe I hadn't tested this, great catch @dimitrimitropulos. I'll look into it, but if you find a solution please let me know.

**Comment by BobTB** [comment]
Edit it outside of the VBE and make make it look like below. It has to have the attribute set, which can not be done in VBE. Use export/import as text. I did it without the #if mac stuff, but it works for me now. 

```
Public Function NewEnum() As IUnknown
Attribute NewEnum.VB_UserMemId = -4
    Set NewEnum = dict_pDictionary.[_NewEnum]
End Function
```

**Comment by retailcoder** [comment]
That requires a `VB_UserMemId = -4` member attribute:

```vb
Public Property Get NewEnum() As IUnknown
Attribute NewEnum.VB_UserMemId = -4
'Gets an enumerator that iterates through the collection.
    Set NewEnum = dict_pKeyValues.[_NewEnum]
End Property
```

...**BUT** I'm using a custom collection class in my current project, and iterating it with a `For Each` loop on a Mac throws "run-time error 57097 / application-defined or object-defined error" - and works perfectly fine on Windows.

IOW I wouldn't classify this issue as a [bug] - it wouldn't work even if it were implemented.

**Comment by skchan2** [comment]
was this bug ever fixed? Having issues with this in the latest version on a Mac

**Comment by dimitropoulos** [comment]
I don't believe it was ever fixed.  I would do so myself, but I have long since moved on from VB.

You could patch it yourself by copying the `cls` file into your project and using it instead, and then add in what I had above and see if you can get it to work.

**Comment by nectorprime** [comment]
Sorry sir

The Converter.bas is okay I can pull data from any REST website, the
problem here is as follows:

*(1) Formatting the Ms Access query into Json string and send data*
*(2) How to use "POST" method*

Sorry I'm not a guru in VBA , that is why I'm struggling like this.

Any help on the above will be highly appreciated

Regards

Chris

**Comment by timhall** [comment]
@skchan2 This hasn't been fixed just yet, I gave it a good go, but couldn't quite get it working right. Will look into it again in the future, but if you get it working please send a PR or a code sample!

**Comment by dimitropoulos** [comment]
it would help a lot if some of this stuff was documented, but my recollection is that it (e.g. `IUnknown`) either isn't documented at all or is documented in such a way that it assumes you already understand it.

If anyone reading this issue knows of documentation for adding `For Each` support on custom data structures, I'm sure it would be welcomed if you mentioned it here.

**Comment by skchan2** [comment]
another thing is that, maybe we can update it so that this class is only referenced when not on a PC (im seeing the for each issue when testing on PC, but run fine when I remove the class).

**Comment by skchan2** [comment]
@dimitropoulos I tried adding your code, but still not able to get it to work. Oddly, the original code is not even working in Windows. Getting stuck in the "For Each" part, weird

`Public Property Get NewEnum() As IUnknown
Attribute NewEnum.VB_UserMemId = -4
    'Gets an enumerator that iterates through the collection.
#If Mac Or Not UseScriptingDictionaryIfAvailable Then
    Set NewEnum = dict_pKeyValues.[_NewEnum]
#Else
    Set NewEnum = dict_pDictionary
#End If
End Property`

**Comment by dimitropoulos** [comment]
yeah, I'm sorry.  I won't be much help to you at this point.  My VB days are behind me and I don't really remember how it works in this case.  What I do remember is that you're stepping into somewhat uncharted territory so my best suggestion is to try and find (or construct) simpler versions of the same functionality for any other collection and see if you can get it to work.   that... or... find some documentation (I was never able to find the documentation for this stuff but that was a few years ago now so maybe it now exists).

**Comment by nectorprime** [comment]
Thank you so much I have finally shifted to dot-net-framework

*Managing Partner*

*Ca Premier Accounting Package Developer - Zambia*
*Nector Prime Accounting Solutions*
*Findeco House 7th Floor Room 8*
*Cairo Road*

*Lusaka, Zambia*

*Nector Prime Accounting Solution is a legally independent firm of
Chartered Accountants and does not accept any responsibility or liability
for the actions or inaction on the part of any other individual
unaffiliated firm or firms.*

P  Please do not print this e-mail unless absolutely necessary. Save our
ENVIRONMENT.


This e-mail may contain privileged and confidential information. If you are
not the intended recipient, you must not copy or distribute it. If you have
received this e-mail in error, please notify the sender and delete the
e-mail immediately. You should ensure that this e-mail and attachment is
virus free. E-mail is not a 100% secure communication medium. We advise you
understand this lack of security and take any necessary measures when
emailing us. We reserve the right to read any email or attachment entering
or leaving our systems from any source without prior notice. Information
about Nector Prime Accounting Solutions can be found at our offices at
Findeco House 7th Floor Room 8 or Call us on Cell No. +260977352401

**Comment by beyphy** [comment]
This is tricky. It's possible that VBA doesn't support such an operation. It's also possible that an additional option or declaration is needed in the .BAS file to make this work.

Based on the last few hours of work, I've come to the conclusion that this operation is likely unsupported. VBA lets you specify default members in a class module. However, it's possible that a default member is not returned in a for-each-next loop. Please see my code samples below starting with the class module:

Option Explicit

'this is in a class module called Test.
'the attribute "Attribute Value.VB_UserMemId = 0" is
'specified in the value function in the
'.BAS file to make value the default member.

Public Function value() As Variant
    Dim temp As Variant
    value = array(1, 2, 3)
End Function

Now onto the tests in a normal module:

Sub test1()
    Dim t As Test
    Dim elem
    Set t = New Test
    For Each elem In t 'runtime error here, object doesn't support property / method
        Debug.Print elem
    Next elem
End Sub

Sub test2()
    Dim t As Test
    Set t = New Test
    Dim elem
    For Each elem In t.value 'prints 1,2,3 with explicit property used
        Debug.Print elem
    Next elem
End Sub

Sub test3()
    Dim t As Test
    Set t = New Test
    Dim elem
    Dim a
    a = t 'assigns a to the default member of Test. This is the value function and it returns an array of 1,2,3
    For Each elem In a 'prints 1,2,3
        Debug.Print elem
    Next elem
End Sub

I also tried using a collection instead of an array in the value function to see if results were any different, but they were not. So based on the tests above, I believe the operation is not supported. But like I said, it's also possible that there's some additional attribute in the .BAS file that needs to be specified to make this work.

**Comment by beyphy** [comment]
I commented yesterday that I didn't think this was supported, but that was incorrect. While I had read how to iterate through custom collections in C# and javascript, I had not done so in VBA. It is a relatively obscure topic. And you will struggle to find the correct sources if you don't know what you're looking for.

The good news is that I have a working version. You can find that version [here](https://github.com/beyphy/VBA_Projects/blob/master/Dictionary.cls). I put it under my own profile for a few reasons I'll get into shortly.

As many people in this post pointed out earlier, to iterate through a custom collection with a for-each-next you need to use NewEnum. And you need to specify `Attribute NewEnum.VB_UserMemId = -4` in the cls file in the NewEnum procedure. The NewEnum procedure can either be a `property get` or a `function`, but it must be public. 

One thing to note about NewEnum is that I believe it only works on collections. I tried it on the array, the dictionary, and the arraylist and none of them worked. While the dictionary implements it, you get an error if you try to use it. You're told that it's either protected or not implemented or something like that. So you can't use a dictionary with NewEnum. That is a problem with your code because it uses a dictionary if the OS is not mac, or if it's available. This is all done through conditional compilation. 

So in order to get this to work, I needed to comment out all the places where you do conditional compilation to work with the native dictionary object. What do I do instead? I created a new function called MakeDictionary. And If the OS is not mac, or if the dictionary is available, I display a messagebox in the constructor letting them know that a dictionary is available and to consider using the MakeDictionary function instead. Since your keys are in the form of an array, I had to iterate through the keys, and add them to the collection that NewEnum uses. I did this in NewEnum procedure.

So overall, I had to make significant changes in your class file. I'm not sure if that's something you'd be interested in, which is why I put it on my own profile. But if that's something you'd be okay with feel free to merge the code (sorry, I don't know how to do a pull request.) I also have no idea if this works on Mac. It should but I don't have a Mac computer to test on.

**Comment by ghost** [comment]
Final update: I believe I was able to get a version with your original code working. As I noted in my earlier post, your code uses a dictionary object when the library is available, or when the OS isn't a mac. This is an issue because NewEnum can't be used on a dictionary and can only be used on a collection. The solution is to iterate through the keys, add them to a collection, and return that collection from NewEnum.

I believe in your code, when it's used a mac or when ScriptingDictionary isn't available, it store the keys in an array. And when it is available, it stores the keys scripting.dictionary object.  So you can do conditional compilation to determine which of these to iterate through, add them to a collection, and return that collection to NewEnum. 

I believe my main changes to your code were adding the NewEnum method, adding a private collection field called Coll at the top of this module, and setting coll = nothing in the class_terminate event. It works for me on Windows, but I don't have a mac to test on so I can't confirm.

**Comment by hrmck** [comment]
> Final update: I believe I was able to get a version with your original code working. Please find the final file [here](https://github.com/beyphy/VBA_Projects/blob/master/Dictionary.cls). 

@beyphy the link is broken, could you update it?

**Comment by ghost** [comment]
@hrmck I deleted the pull request. But you should be able to find a link to my updated code [here](https://github.com/beyphy/VBA-Dictionary/blob/master/Dictionary.cls)

**Comment by Greedquest** [comment]
This could be achieved by constructing the IEnumVariant struct in memory e.g.: https://stackoverflow.com/a/52261687/6609896

`Scripting.Dictionary` does not expose [_NewEnum] like `Collection` does, nor does the array returned by `dict.keys()`. Therefore there is no way to "forward" for..each calls onto the underlying type - unless it's a collection - so you have to make your own structure to  do the enumeration (like in the linked question).

**Comment by SHIMADONBEY** [comment]
@skchan2 @dimitropoulos made a modification based on the code presented.
The following code was able to verify that it works with the Windows version of Excel VBA. Unfortunately, I do not have a Mac, so I have not been able to verify the operation of the Mac version of Excel VBA.

```VBA
Public Function NewEnum()
Attribute NewEnum.VB_UserMemId = -4
    'Gets an enumerator that iterates through the collection.
#If Mac Or Not UseScriptingDictionaryIfAvailable Then
    Set NewEnum = dict_pKeyValues.[_NewEnum]
#Else
    Set NewEnum = CallByName(dict_pDictionary, "_NewEnum", vbMethod)
#End If
End Function
```
I wasn't sure if the IUnknown interface would be supported by the Mac version of VBA, so I changed the return value to Variant. (This has been confirmed to work in my environment.)

**Comment by Nick-vanGemeren** [comment]
>`    ' Run-time error '438': Object doesn't support this property or method`
>`    For Each z In x`
While it doesn't replicate the Scripting functionality, this 'feature' does force you to consider whether you want to iterate on Keys or Items.

`    For Each z In x.Keys `
... works fine and is clearer code.

**Comment by DecimalTurn** [comment]
Here's how [VBA-FastDictionary](https://github.com/cristianbuse/VBA-FastDictionary) does it : https://github.com/cristianbuse/VBA-FastDictionary/blob/028e151af3436e71802bc6fecf9676fc1e7e3727/Implementation.md#newenum
